### DISCUSSION TITLE
Error when process a large file

### DISCUSSION THREAD

**Discussion by IbrahimAlhaj9** [nextgenhealthcare/connect#5189] [General]
Hi Everyone.
When I am try to process a file exceed 25 MB , the following error is occur :
```
Filter error
ERROR MESSAGE: Error evaluating filter
java.lang.OutOfMemoryError: Java heap space
```

I think the issue is coming from this function:
```
msg = new XML(SerializerFactory.getSerializer('HL7V2').toXML(Message));
```

Can you please help me it's urgent for me

Thank you in advance

**Comment by FTK7** [comment]
Have you tried to increase Server Memory? Not sure if it could help.
![imatge](https://user-images.githubusercontent.com/66680696/168258838-98f3f17a-c557-4b85-b411-c4e7f7f78f1e.png)

**Reply by IbrahimAlhaj9** [reply]
@FTK7 the server memory is 1024-Thanks

**Comment by odoodo** [comment]
You probably use a large inline attachment in your hl7 message. Use an attachment handler to avoid this issue. More details can be found [here](https://forums.mirthproject.io/articles/175046-using-attachments).

**Reply by IbrahimAlhaj9** [reply]
@odoodo I am trying the Attachments Using JavaScript solution:
input HL7 Listener
Output: Javascript writer
I write the following code : 

```
var attach = addAttachment(msg, "text/plain", false)
return attach.getAttachmentId();
```

then I need to do the following:
before the destination sends the message out, it will replace the attachment id with the attachment.

How i can do that?
currently i have the following code in the destination:
```
var FileName = $('FileName');

var encData = FileUtil.encode(new java.lang.String(SerializerFactory.getSerializer('HL7V2').fromXML(connectorMessage.getEncodedData())).getBytes())

var ret = new String('<Message>')
var datestamp=DateUtil.getCurrentDate('yyyyMMddHHmmssSSS') 

ret+='<Body FileName="'+ FileName + '.pdf' + '" Type="hl7" Include="true">' +  encData + '</Body>'


ret+='<Files>'


ret+='</Files>'

ret+='</Message>'

$c('chanretChannelMessage','Success')
$c('chanretLastChannelStep','Destination Writer')
$c('chanretStatusCode','200')
return XmlUtil.prettyPrint(ret)
```

Thank you in advance

**Reply by odoodo** [reply]
First of all we would need to know if you are treating an HL7 message containing an attachement (like indicated by your first post) or if the whole message is the attachment (indicated by your last post).

Depending on your destination connector, the attachment will be reattached if you are enabling the "Reattach Attachment"-option.
But also here, we would have to know your setting. What kind of destination connector are you using?

**Reply by IbrahimAlhaj9** [reply]
I am trait a message contains pdf data as base64 in obx segments
I try the solution mentioned by you in the following link:
https://forums.mirthproject.io/articles/175046-using-attachments
but I didn't know how.

**Reply by IbrahimAlhaj9** [reply]
![source](https://user-images.githubusercontent.com/54806435/168305452-d561e851-64a1-4aad-a5d9-6f81afd03780.PNG)
![destination](https://user-images.githubusercontent.com/54806435/168305468-53f4caa8-5d8b-4136-b7eb-21ccaf6bdddc.PNG)

**Reply by odoodo** [reply]
In the summary tab of your channel configure the following:
![AttachmentConfig_1](https://user-images.githubusercontent.com/61003874/168306196-41f567c2-8152-43a6-8970-918310d438b7.PNG)
(Assuming, the pdf is stored in OBX-5.5)

In your destination connector configure the following:
![AttachmentConfig_2](https://user-images.githubusercontent.com/61003874/168306202-02c55b96-cfe3-4005-8884-9bc071cc079d.PNG)
(in your case its a javascript writer but the option is the same)

Now, if you place the attachment id somewhere in the code of your javascript writer, it will automatically be replaced by the actual attachment before sending. (what you currently try to do with encData)

For setting the attachment id, drag&drop attachmentId from the right panel or insert **$('attachmentId')** at the right spot in your code.

**Reply by IbrahimAlhaj9** [reply]
I defined the attachment expression like the first image and I changed the destination like the following:
![New](https://user-images.githubusercontent.com/54806435/168530481-967e2b5a-2881-4d9f-9262-d1a6fc0990c5.PNG)
Result:
OBX|8|ED|PDFReport1^PDFReport||Novo^Image^PDF^Base64^${ATTACH:1019fc6b-4938-40c4-ae4e-4aca7226076f}||||||C

what you currently try to do with encData?
encData contains the translated message as base64 :it send to my website through API and store into database

How I can keep use the encData  with the attachments? and send to my website the message with the attachment?

Thank you

**Reply by odoodo** [reply]
> OBX|8|ED|PDFReport1^PDFReport||Novo^Image^PDF^Base64^${ATTACH:1019fc6b-4938-40c4-ae4e-4aca7226076f}||||||C

Looks good. Mirth should replace the reference by the actual attachment when sending the message

> what you currently try to do with encData?

I thought that was the var into which you where you were trying to load the attachment. If not, ignore.

> How I can keep use the encData with the attachments?

I am not sure what you mean by that

**Reply by IbrahimAlhaj9** [reply]
@odoodo I mean by:
``
How I can keep use the encData with the attachments?
``
the output message wasn't replaced by the pdf data.
I debug the code and I found that the following function has the issue:
``
connectorMessage.getEncodedData()
``
this function return the pdf as following:
**OBX|8|ED|PDFReport1^PDFReport||Novo^Image^PDF^Base64^${ATTACH:1019fc6b-4938-40c4-ae4e-4aca7226076f}||||||C**

how I can fix this issue? I want the pdf data to be inside the encData variable 
thank you

**Reply by odoodo** [reply]
I am still not sure, if I understand correctly what you are trying to do but try replacing line 3 by 

```javascript
var encData = connectorMessage.getEncodedData();
```

**Reply by IbrahimAlhaj9** [reply]
My issue is at the following function:
```
connectorMessage.getEncodedData();
```
it didn't replaced automatically the pdf data when try to print the data like the following:
```
logger.info(connectorMessage.getEncodedData());
```

current output:
```
OBX|8|ED|PDFReport1^PDFReport||Novo^Image^PDF^Base64^${ATTACH:1019fc6b-4938-40c4-ae4e-4aca7226076f}||||||C
```

Expected output:
```
OBX|8|ED|^PDF Report^PDFReport1^^L||^^PDF^Base64^JVBERi0xLjUNCjQgMCBvYmoNCjw8L1R5cGUgL1BhZ2UvUGFyZW50IDMgMCBSL0NvbnRlbnRzIDUgMCBSL01lZGlhQm94IFswIDAgNjEyIDc5Ml0vUmVzb3VyY2VzPDwvRm9udDw8L0ZBQUFBSSA4IDAgUi9GQUFBQkIgMTEgMCBSPj4vWE9iamVjdDw8L1gxIDYgMCBSPj4+Pi9Hcm91cCA8PC9UeXBlL0dyb3VwL1MvVHJhbnNwYXJlbmN5L0NTL0RldmljZVJHQj4+Pj4NCmVuZG9iag0KNSAwIG9iag0KPDwvTGVuZ3RoIDEzIDAgUi9GaWx0ZXIgL0ZsYXRlRGVjb2RlPj5zdHJlYW0NCnic7VpZd9u4Ff4rfEznVDD2JW+yrNhMrSUSp530pUdRFFutJWUkTabz73sBkCLAZSyP5SU9Iyc2eUEAH+5+L0USDD8dAr+Uocl8lfycEEcjCWeIYvtRMiHajnVCEk5WiZAaKWOMVsoA4S4mKI2MfZgIBkPR5HjoNvlHsg52hv+IcGM/2mMChPBvlVCuEbcj0ghLuYspym1E/N+mVYVfjhLpDy4tbPgF1LOfSHKxST64n3IW5apcHzsuFFiU8SiZyaEEhAYk51ly9q4LnzQhPMm+5FvAQ4RqhKkEfsB5iUZMCViFiyRbwQM3yZvh5ttmPNvf/iX7d/Imcb/t7eZuc/NbQLuefbJ/+1l8aomwsdw2zMIPYZAYhiJIM3tUxhMikFDCSwtgvGGYBDsNUXBzNdtul7vN2t4ogbiEzQyHNbPPxSPT/Xax2NvLjuSIKI4JoQw2pUhyi42oJLuAU22X6/liv1n/NdzsfXCDteA4P+WH6JycIAZSAvbL+86JkSHAE8F17Zzj28168TbYkHRAmTtS6o4EyI07C44wB/ERfh+HNUWCwnmlkbWd383+6/alcA5NjNZgGyEPIyD8AORDBYxiiIjctABMPBzZIOOIYbglRvwB66YCUaPhaVCWinXHQ8dbt6ASWc1XQuQmFVFAWbjRVj+9bUX3fpeYFYQcHjjxPoWIz88TgiMRU44RkdaWmUoMYsxecn4w5fHsZhFqV3C9+VIdaBQ
```

**Reply by odoodo** [reply]
```javascript
var encData = AttachmentUtil.reAttachMessage(connectorMessage);
```

**Reply by pacmano1** [reply]
You should post a full export of your channel.

**Reply by IbrahimAlhaj9** [reply]
@odoodo , @pacmano1 
[Channel-with-attachment.zip](https://github.com/nextgenhealthcare/connect/files/8708201/Channel-with-attachment.zip)

**Reply by odoodo** [reply]
Did you try the proposal from my last post?

**Reply by IbrahimAlhaj9** [reply]
No sorry I didn't notice , I will try it

**Reply by IbrahimAlhaj9** [reply]
Thank you very much it's working Now

**Reply by pacmano1** [reply]
I am puzzled the attachment works, the inbound message type is XML, not HL7.  I would not expect the attachment on its own to exist given it is a regex collection.

**Reply by odoodo** [reply]
The attachment is extracted from the raw message before it gets parsed into the channel. (Or do you mean the inbound message itself is xml?)

**Reply by IbrahimAlhaj9** [reply]
@pacmano1  , @odoodo 
yes the input is Xml but i transform it to HL7 inside Filter as follows:
```
function extractMultiPart() {	
	var AttachmentsFileNames =[];
	var AttachmentsContentTypes =[];
	var AttachmentsNames = [];
	var AttachmentsContents = [];
	var Message;
	var MessageFileName;
	var PDFReport='';
	var PDFReportFileName='';
	var PreTransformFunctions='';
	var PostTransformFunctions='';
	var SendingFacilities='';
	var ReceivingFacilities='';
	
	for (var i = 0; i < getArrayOrXmlLength(msg['Content']['Part']); i++) {
	    
	    var disposition =msg['Content']['Part'][i]['Headers']['Content-Disposition']
	    var filename='';
	    var name='';	
	    
	    if (disposition && disposition.indexOf(' filename') !== -1) {
	        var filenameRegex = /filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/;
	        var matches = filenameRegex.exec(disposition);
	        if (matches != null && matches[1]) { 
	          filename = matches[1].replace(/['"]/g, '');
	        }
	    }    
	    if (disposition && disposition.indexOf(' name') !== -1) {    
	        var filenameRegex = /name[^;=\n]*=((['"]).*?\2|[^;\n]*)/;
	        var matches = filenameRegex.exec(disposition);
	        if (matches != null && matches[1]) { 
	          name = matches[1].replace(/['"]/g, '');
	          name=name.toLowerCase()
	        }        
	    }
	    
	    
	    if (name=='message'){
			if (msg['Content']['Part'][i]['Content']['@encoding']=='Base64') 
				Message=new java.lang.String(FileUtil.decode(msg['Content']['Part'][i]['Content'].toString().trim()));
			else
				Message=new java.lang.String(msg['Content']['Part'][i]['Content'].toString().trim());
			
			Message=Message.replaceAll("^\u000B", "")
			Message=Message.replaceAll("\u0016$", "")
			Message=Message.replaceAll("\u001C$", "")
				
			if (filename !='') MessageFileName=filename
	    }
	    else if (name=='messagefilename'){
			MessageFileName=msg['Content']['Part'][i]['Content'].toString().trim();    	
	    }	    
	    else if (name=='pdfreport'){
			PDFReport=msg['Content']['Part'][i]['Content'].toString().trim();    	
			if (filename !='') PDFReportFileName=filename
	    }
	    else if (name=='pdfreportfilename'){
			PDFReportFileName=msg['Content']['Part'][i]['Content'].toString().trim();    	
	    }		    
	    else if (name=='pretransformfunctions'){
			PreTransformFunctions=msg['Content']['Part'][i]['Content'].toString();    	
	    }
	    else if (name=='posttransformfunctions'){
			PostTransformFunctions=msg['Content']['Part'][i]['Content'].toString();    	
	    }
	    else if (name=='sf'){
			SendingFacilities=msg['Content']['Part'][i]['Content'].toString();    	
	    }
	    else if (name=='rf'){
			ReceivingFacilities=msg['Content']['Part'][i]['Content'].toString();    	
	    }		    	
	    else {
	    		AttachmentsNames.push(name);
		
		    	AttachmentsFileNames.push(filename);
		
			var ContentType = msg['Content']['Part'][i]['Headers']['Content-Type'].toString();
		
			if (ContentType == null)
				ContentType=''

			AttachmentsContentTypes.push(ContentType);
			
			var Content = msg['Content']['Part'][i]['Content'].toString();
			
			if (Content == null)
				Content=''
							
			AttachmentsContents.push(Content);
			
			
	    }

	
	}

	$c('AttachmentsNames',AttachmentsNames);
	$c('AttachmentsFileNames',AttachmentsFileNames);
	$c('AttachmentsContentTypes',AttachmentsContentTypes);
	$c('AttachmentsContents',AttachmentsContents);
	$c('Message',Message);
	$c('MessageFileName',MessageFileName);
	$c('PDFReport',PDFReport);
	$c('PDFReportFileName',PDFReportFileName);
	$c('PreTransformFunctions',PreTransformFunctions);
	$c('PostTransformFunctions',PostTransformFunctions);
	$c('ReceivingFacilities',ReceivingFacilities);
	$c('SendingFacilities',SendingFacilities);
	
	try
	{
		if (Message.trim()=='')
			msg=new XML('')
		else if(Message.substring(0,3)=='MSH')
			msg = new XML(SerializerFactory.getSerializer('HL7V2').toXML(Message));
		else
			msg=new XML(Message)
	}
	catch(err){
		logger.error('Error parsing message in extractMultiPart: '+ Message)
		msg=new XML('')
	}
		
}
```
What do you think about Attachments? and How I can configure attachments with xml ,json ?

**Reply by pacmano1** [reply]
Read the article again on attachments.   That's just another regex exercise.
